Merge HAFSv2.1.1 real-time parallel experiment related developments back to develop

BinLiu-NOAA opened the pull request:
## Description of changes
Provide a description of what this PR does. What bug does it fix, or what feature does it add? Do you expect that this PR will change answers, and if so, under what circumstances? If this PR is for a physics innovation, please provide references to any relevant scientific papers.

Merge HAFSv2.1.1 real-time parallel experiment related developments back to HAFS develop branch.

## Issues addressed (optional)
If this PR addresses one or more issues, please provide link(s) to the issue(s) here.
- fixes hafs-community/HAFS/issues/318

## Contributors (optional)
If others worked on this PR besides the author, please include their user names here (using @Mention if possible).

## Tests conducted
What testing has been conducted on the PR thus far? Describe the nature of any scientific or technical tests, including relevant details about the configuration(s) (e.g., cold versus warm start, number of cycles, forecast length, whether data assimilation was performed, etc). What platform(s) were used for testing?

Extensive technical and real-time testing has been conducted for the 2025 hurricane season.

## Application-level regression test status
Running the HAFS application-level regression tests is currently performed by code reviewers after the developer creates the initial PR. As regression tests are conducted, the testers should use the checklist below to indicate **successful** regression tests. You may add other tests as needed. If a test fails, do not check the box. Instead, describe the failure in the PR comments, noting the platform where the test failed.

- [x] Ursa
- [x] Gaeac6
- [x] Orion
- [x] Hercules
- [x] WCOSS2

wramstrom committed:
Added links for static surface data files for NOAHMP nest move.

wramstrom committed:
Updates to NOAHMP For moving nest to correct land sea mask.

wramstrom committed:
Changed default value for soilcolor to 10, based on advice from Mike Barlage.

wramstrom committed:
Add moving nest if condition to NOAHMP changes

wramstrom committed:
Merge remote-tracking branch 'remotes/origin/feature/hafsv2_phase2' into feature/mn_noahmp

wramstrom committed:
Restructured moving nest fix file initialization.

wramstrom committed:
Removed trailing whitespace.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs.v2.0.1' into feature/mn_noahmp

BinLiu-NOAA committed:
Add example conf and cronjob files for using NoahMP LSM.

BinLiu-NOAA committed:
* Update submodule sorc/hafs_forecast.fd, which synced with latest
  ufs-weather-model as of 06/28/2025, and including the scale-aware
  3DTKE EDMF GFS PBL scheme related changes from Prof. Ping Zhu (@zhup01)
  from FIU and and @samuelkyfung
* Add related namelist and config options of sa3dtke_dyco and sa3dtke.

RongqianYang-NOAA committed:
Update .gitmodules to own fork

Commit:
Change UFS to feature branch

Commit:
Update .gitmodules in HAFS

BijuThomas-NOAA committed:
Porting HAFS(sub components) to Ursa

BijuThomas-NOAA committed:
Updates in hafs_forecast.fd component for Ursa porting

BijuThomas-NOAA committed:
Porting HAFS to Ursa

BijuThomas-NOAA committed:
Porting HAFS to Ursa(Workflow related changes)

BijuThomas-NOAA committed:
Adding Ursa for loading platform specific modules for Analysis jobs

BijuThomas-NOAA committed:
Enable EMC Graphics to Ursa platform

BijuThomas-NOAA committed:
Porting HAFS to Ursa(updates in hafs_forecast.fd)

BijuThomas-NOAA committed:
Porting HAFS to Ursa(Update "COMrtofs" in system.conf.ursa)

BijuThomas-NOAA committed:
Porting HAFS to real Ursa

BijuThomas-NOAA committed:
Porting HAFS to real Ursa

BijuThomas-NOAA committed:
Pointing the updates in hafs_tracker.fd(Porting HAFS to real ursa)

BijuThomas-NOAA committed:
Porting HAFS to Ursa(Updated in system.conf.ursa & ursa.ent)

BinLiu-NOAA committed:
Sync submodules of sorc/hafs_forecast.fd (as of 04/25/2025) and
sorc/hafs_gsi.fd (as of 05/06/2025) with their latest develop branch.

BijuThomas-NOAA committed:
Pointing to the updates in the "hafs_gsi.fd" submodule

BinLiu-NOAA committed:
Update submodule sorc/hafs_utils.fd.

BijuThomas-NOAA committed:
Pointing to the updates in the "hafs_gsi.fd" submodule

BijuThomas-NOAA committed:
Update the Ursa partition name for accessing general compute  resources

BijuThomas-NOAA committed:
Syncing hafs_gsi.fd and hafs_tracker.fd with remotes/origin/production/hafs.v2.1

BijuThomas-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafsv2p1_final' into feature/hafsv2p1_ursa

BijuThomas-NOAA committed:
Pointing the new submodule level commit in hafs_gsi.fd

BinLiu-NOAA committed:
* Update submodule hafs_forecast.fd (turning off the ENABLE_PARALLELRESTART build option on WCOSS2)
* Swithch to use the previous version of hafs_utils.fd
* Update FV3 input nml and MOM6 input file

BijuThomas-NOAA committed:
Use srun to execute hafs_mom6_obc_from_rtofs.py that has ESMF/esmpy
utilities

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafsv2p1_final' into feature/hafs.v2.1.1

BijuThomas-NOAA committed:
Pointing to submodule level update(hafs_gsi.fd)

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafsv2p1_final' into feature/hafs.v2.1.1

BinLiu-NOAA committed:
Update submodule hafs_forecast.fd, which was synced with ufs-weather-model develop branch as of 06/02/2025.

BijuThomas-NOAA committed:
Fixing the logic to source storm1.holdvars.txt file before ocean_model
variable is used in JHAFS_OCN_PREP

XuLu-NOAA committed:
Fix the HAFS ensemble capability with dual-reso ensemble DA.

BijuThomas-NOAA committed:
Changing the directory location from /scratch3/NCEPDEV to /scratch3/HFIP and
 disk_project/cpu_account name to hwrfv3 in "system.conf.ursa"

BijuThomas-NOAA committed:
Pointing to the updated commit at submodule(sorc/hafs_forecast.fd)

binli2337 committed:
Update exhafs_forecast.sh.

BinLiu-NOAA committed:
Properly support coupled HAFS forecast dealing with WW3 point output in NetCDF format.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/develop' into feature/hafs.v2.1.1

BinLiu-NOAA committed:
Update submodule sorc/hafs_forecast.fd (synced with ufs-weather-model develop as of 06/12/2025)

BinLiu-NOAA committed:
Merge remote-tracking branch 'hcomm/feature/hafs.v2.1.1' into feature/EMC_mn_noahmp

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/develop' into feature/hafsv2p1_ursa

BinLiu-NOAA committed:
Update link_fix.sh to use correct fix file location on Ursa.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafsv2p1_ursa' into feature/hafs.v2.1.1

BinLiu-NOAA committed:
Merge remote-tracking branch 'origin/feature/hafsv2p1_ursa' into feature/hafs.v2.1.1

BinLiu-NOAA committed:
Update k_split and n_split for parm/rt_conf/hafs_globnest_atm.conf and parm/rt_conf/hafs_globnest_static_atm.conf.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs.v2.1.1' into feature/EMC_mn_noahmp

BinLiu-NOAA committed:
Update submodule sorc/hafs_gsi.fd.

BinLiu-NOAA committed:
Merge remote-tracking branch 'origin/feature/hafs.v2.1.1' into feature/EMC_mn_noahmp

BinLiu-NOAA committed:
Update sorc/hafs_gsi.fd for Gaeac6 build.

XuLu-NOAA committed:
Merge remote-tracking branch 'origin/feature/hafs.v2.1.1' into feature/hafsv2p1_ursa_hafsens

BijuThomas-NOAA committed:
Pointing the submodule update(hafs_gsi.fd)

BijuThomas-NOAA committed:
Updates in submodule(hafs_gsi.fd)

BinLiu-NOAA committed:
Merge remote-tracking branch 'origin/feature/hafs.v2.1.1' into feature/EMC_mn_noahmp

BinLiu-NOAA committed:
Merge remote-tracking branch 'origin/feature/hafs.v2.1.1' into feature/hafsv2p1_ursa_hafsens

BinLiu-NOAA committed:
Changes to run ocn_post job after forecast job completes.

BinLiu-NOAA committed:
Changes to improve configuring and running HAFS coldstart ensembles for ENSDA.

BinLiu-NOAA committed:
Add example config and cronjob driver for the hafs.v2.1.1a ensda experiment.

BinLiu-NOAA committed:
Add default all_tasks_ens in forecst_ens section in hafs.conf.

BinLiu-NOAA committed:
Further improved using ncks --deflate=1 in exhafs_atm_post.sh and exhafs_merge.sh.
Note: for the analysis/gsi job, it cannot handle some netcdf files
(e.g., grid_spec, akbk files) after the ncks --deflate=0 compression.

BinLiu-NOAA committed:
- Update default atm_init PE layout settings
 - Update and unify HAFS supported platforms' site entity files
 - Enable properly setting nsst related options for HAFS ensemble forecast

BinLiu-NOAA committed:
Properly configure and turn on an experiment configuration of
HAFSv2.1.1A 4DEnVar with mixed 40 HAFS ensembles and 80 GDAS ensembles.

BijuThomas-NOAA committed:
Fix in the application level module file and, pointing to the submodule level commits(Tracker and Utils), after the new CPE on Gaea C6

MariaAristizabal-NOAA committed:
Clean the MOM_input files by removing the unused flags. In addition, MEKE was set to False as recommended by GFDL.

BinLiu-NOAA committed:
Update sorc/hafs_forecast.fd and enable configuring and using GFDL MPv3.

BinLiu-NOAA committed:
Add hafsv2p1p1a_baseline and hafsv2p1p1a_gfdlmpv3 conf files and cronjob drivers.

BinLiu-NOAA committed:
Update submodule scripts/exhafs_forecast.sh, which enables wave-ocean coupling capability.
Enable using cpl_wav_ocn to control wave-ocean coupling
From @binli2337, @MariaAristizabal-NOAA, @BijuThomas-NOAA, @BinLiu-NOAA

BinLiu-NOAA committed:
Update submodule sorc/hafs_gsi.fd and turn on outputting observation
space ensemble spread in GSI diag files. (from @DanWu-HRD)

BijuThomas-NOAA committed:
Add "hafs_calc_analysis" to the HAFS build system(within "hafs_tools.fd"
component) and corresponding updates in modulefiles

BijuThomas-NOAA committed:
Merge branch 'feature/hafs.v2.1.1' of https://github.com/hafs-community/HAFS into feature/hafs.v2.1.1

JunghoonShin-NOAA committed:
1. Modified exhafs_atm_vi.sh, anl_pert and anl_combine to add the HMON type 9-point smoother (B-grid),2. Modified split.90 to fix the run time issue during the reading input files, 3. Add hmon_vi options in hafs_holdvars.txt, hafs_holdvars.conf, hafs.conf

BijuThomas-NOAA committed:
Loading correct version of ncio on WCOSS2 in the modulefile

JunghoonShin-NOAA committed:
Upudate and improve the script and config files for the new VI updates

BinLiu-NOAA committed:
Update hafsv2p1p1a_baseline and hafsv2p1p1a_gfdlmpv3 conf files and cronjob drivers.

BinLiu-NOAA committed:
Remove duplicated module lines in hafs.gaeac6.lua and hafs.wcoss2.lua.

BinLiu-NOAA committed:
Update build_all.sh to build subcomponents simultaneously in the backend.

BijuThomas-NOAA committed:
Update in calc_analysis: Integrated the ncio.f90 module with the calc_analysis source files to ensure both use the same version of NetCDF

BinLiu-NOAA committed:
Update scripts/exhafs_atm_ic.sh so that it calls
hafs_tools_calc_analysis.x to calculate enkfgdas ensemble member
analysis to be used for HAFS coldstart ensemble member initialization.

BinLiu-NOAA committed:
Update submodule sorc/hafs_forecast.fd, which got synced with UWM's PR
of GFDLMP-v3 and TEMPO MP microphysics #2503.

BinLiu-NOAA committed:
Revert adding hafs_calc_analysis to the HAFS build system within hafs_tools.fd.

This reverts the following commits:
7562688edf35c5c08df32970f11f9dfe72133eae
d34dd3c9cfc3a949567b3c3642fa481818191319
4eed303f4ec70b95af3f1524bb7c6000b38c1262
0414ed9ab5d221306caa106bb10f327484727f93
ace0b89410c39044834f45d236b765140f8a4708

BinLiu-NOAA committed:
* Enable initializing HAFS DA ensemble members from ENKFGDAS ensemble
  analysis instead of 6-hour forecast.
* The ush/hafs_calc_enkfgdas_analysis.py is added to generate the analysis
  based on 6-hour forecast/guess and the analysis increment file.

BijuThomas-NOAA committed:
Updating Modules for Hera; pointing new commits in submodules(hafs_gsi.fd, hafs_tracker.fd, hafs_utils.fd)

BijuThomas-NOAA committed:
Removed the --exclusive flag from Slurm job directives—except for the forecast job—to increase core availability on Ursa nodes

BinLiu-NOAA committed:
Adding the TDR observation superobbing capability (#314)

* Add the hafs_tdr_superob source code (From @Brittany-Dahl)
* Adding hafs_tdr_superob to HAFS build system (From @BijuThomas-NOAA)
* Add a config option of tdr_superob and related script changes to enable
   TDR superobbing. Update submodule sorc/hafs_gsi.fd accordingly. (From @Brittany-Dahl)

Co-authored-by: @Brittany-Dahl, @BijuThomas-NOAA

BinLiu-NOAA committed:
Update submodule hafs_forecast.fd, which has 3DTKE GFS PBL related fixes.

BijuThomas-NOAA committed:
Loading "py-netcdf4" and "py-netcdf4" in hafs.hera.lua(Hera platform)

Commit:
Set the bug related flags in MOM_input to the value for which the corrected code is
used. The exception to this is the flag USE_HUYNH_STENCIL_BUG, which is
set to True so MOM6 is reproducible on WCOSS2 (recommende by Jiande
Wang).

BijuThomas-NOAA committed:
Adding "numactl --interleave=all" command on Ursa to set a NUMA memory allocation policy that distributes memory evenly across all NUMA nodes

BinLiu-NOAA committed:
Update submodule hafs_forecast.fd.

BinLiu-NOAA committed:
The following are needed for using NoahMP LSM.
Turn off sfc_data file merge in scripts/exhafs_merge.sh.
Turn on link sfc_data file in scripts/exhafs_forecast.sh.
From @RongqianYang-NOAA and @ChuanKaiWang-NOAA

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs.v2.1.1' into feature/EMC_mn_noahmp

BinLiu-NOAA committed:
Rename and update parm/hafsv2p1p1a_noahmp.conf and rocoto/cronjob_hafsv2p1p1a_noahmp.sh

BinLiu-NOAA committed:
From @RongqianYang-NOAA: Change input nml option of iopt_dveg from 4 to
1 when using GFSv16 input data. In the future, when using GFSv17 input
data (which uses NoahMP LSM), will change this option back to 4.

BinLiu-NOAA committed:
Enable using TTE-EDMF option for GFS PBL scheme developed by @JongilHan66.
Besides, the tte_edmf option can also be combined with the 3DTKE option.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/develop' into feature/hafs.v2.1.1

BijuThomas-NOAA committed:
Fixing a typo in hafs_runcmd.sh.inc(Run command on Ursa)

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/MOM_input_fix' into feature/hafs.v2.1.1

BinLiu-NOAA committed:
Enable TDR superobbing working properly under both production and retrospective modes.

JunghoonShin-NOAA committed:
Modified config and namelist files for the GFDL MPv3

BijuThomas-NOAA committed:
Fixing a typo in sites/ursa.ent file

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/EMC_mn_noahmp' into feature/hafs.v2.1.1

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafsv211_tte_edmf' into feature/hafs.v2.1.1

BinLiu-NOAA committed:
Update submodule sorc/hafs_tracker.fd.

BinLiu-NOAA committed:
Update parm/hafsv2p1p1a_baseline.conf.
Add hafsv2p1p1a_final conf file and cronjob driver.

BinLiu-NOAA committed:
Clean up hafsv2p1p1a noahmp, gfdlmpv3, mixensda conf files and cronjob drivers.

BinLiu-NOAA committed:
Turn off write_obs_sprd in parm/analysis/gsi/gsiparm.anl.tmp before
being fixed to work with regional HAFS ensembles.

BinLiu-NOAA committed:
Update submodule sorc/hafs_gsi.fd, containing latest S2 UAS HDOB
observation related change from @XuLu-NOAA.

BinLiu-NOAA committed:
Update submodule hafs_forecast.fd, which has 3DTKE related updates from @JongilHan66.

JunghoonShin-NOAA committed:
Update submodule hafs_forecast.fd for the modified Tiedtke convection scheme

BijuThomas-NOAA committed:
Fixing the order of sources in CMakeLists.txt for hafs_tdr_superob(from yonghui.weng@noaa.gov)

BinLiu-NOAA committed:
Update parm/hafsv2p1p1a_final.conf by setting tdr_superob=.true. and cpl_wav_ocn=cmeps_1way_2to1.

BinLiu-NOAA committed:
Update hafsv2p1p1a_final.conf to not run the wave component (WW3).

BinLiu-NOAA committed:
Script changes to enable saving a copy on disk for HAFS ensemble restart files.
Note: In the future, these changes might not be needed, thus can be reverted.

BinLiu-NOAA committed:
Update system.conf.ursa:CDNOSCRUB=/scratch3/HFIP/{disk_project}/noscrub/{ENV[USER]}/hafstrak

BinLiu-NOAA committed:
Update exhafs_obs_prep.sh so that it checks and uses both regular and S2 UAS HDOB observations (if existing).

BinLiu-NOAA committed:
Update exhafs_archive.py so that it can properly handle the scenario of
no HAFS ensemble restart files for the ens_disk archive option.

BinLiu-NOAA committed:
From @Brittany-Dahl: Update hafs_tdr_superob/module_tdr_superob.f90 by
adding a check that forces the sweep to change if the number of radials in
the sweep exceeds max_radials_swp, even if a large change in azimuth
angle has not been detected, in the "split_bufr_sweeps" subroutine.

BinLiu-NOAA committed:
Update hafs_runcmd.sh.inc so that it adds numactl --interleave=all for srun
commands on Hera (like for Ursa). This will help avoid the forecast job
failures due to lack of memory.

BinLiu-NOAA committed:
Revert "Update hafs_runcmd.sh.inc so that it adds numactl --interleave=all for srun"

This reverts commit 25dd6a7b9b3e19d6231ceb32ee0b22820ea375d6.

BinLiu-NOAA committed:
Update submodule sorc/hafs_gsi.fd, which includes a fix for assimilating
high-resolution AMVs (satwhr) for WV, VIS, and OP.

BijuThomas-NOAA committed:
Updating Platform identification on Hera after scratch[1/2] were decommissioned

JunghoonShin-NOAA committed:
@JunghoonShin-NOAA @ChuankaiWang-NOAA @BinLiu-NOAA
Changing split.f90, anl_combine.f90, storminfo.py to fix the VI failure near IDL

BinLiu-NOAA committed:
Update submodule sorc/hafs_forecast.fd.

BinLiu-NOAA committed:
Update and finalize hafsv2p1p1a_final.conf.

BinLiu-NOAA committed:
Add hafsv2p1p1a_jtwc.conf and cronjob_hafsv2p1p1a_2025rt.sh.

BinLiu-NOAA committed:
Update PE layout for hafsv2p1p1a_final.conf and hafsv2p1p1a_jtwc.conf.

BinLiu-NOAA committed:
Update rocoto/sites/hera.ent to use OMP of 4 for the forecast job.

BinLiu-NOAA committed:
Update hafsv2p1p1a_final.conf and hafsv2p1p1a_jtwc.conf to only use 60 PEs for ocean coupling.

BinLiu-NOAA committed:
Update scripts/exhafs_obs_prep.sh to properly handle S2 UAS HDOB observations.
Also, for tdr_superob, check exist and non-empty for ./tldplr.ibm before
sending to intercom.

BijuThomas-NOAA committed:
Pointing to the submodule level(hafs_graphics.fd/emc_graphics) update

BijuThomas-NOAA committed:
Pointing to the submodule level (hafs_forecast.fd: Stack update on Gaea C6)

BinLiu-NOAA committed:
Update submodule sorc/hafs_forecast.fd with NoahMP moving nesting related updates/fixes.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs.v2.1.1_gaeac6' into feature/hafs.v2.1.1

BinLiu-NOAA committed:
Update .gitmodules.

BinLiu-NOAA commented:
@JunghoonShin-NOAA, @BijuThomas-NOAA, @ChuankaiWang-NOAA, @LinZhu-NOAA, Please help to conduct the regression tests for this PR for the following platforms, respectively.
 - Ursa: @JunghoonShin-NOAA 
 - Gaeac6: @BijuThomas-NOAA 
 - Orion: @ChuankaiWang-NOAA 
 - Hercules: @ChuankaiWang-NOAA 
 - WCOSS2: @LinZhu-NOAA 
 
Thanks!

LinZhu-NOAA commented:
The analysis job failed for hafs.v2.1.1 for regional_static_C192s1n4_atm_3denvar on WCOSS2. The error message is as follows:
ncdf error NetCDF: HDF error. 
ncdf error NetCDF: HDF error. 
ncdf error NetCDF: HDF error. 

I rewinded the job, it still failed. The log file is as follows:
/lfs/h2/emc/ptmp/lin.l.zhu/hafs.v2.1.1_rt_regional_static_C192s1n4_atm_3denvar/2024092412/00L/hafs_analysis.log 
Thanks!

ChuankaiWang-NOAA commented:
I have this error message for regional_static_C192s1n4_atm_3denvar on Hercules and Orion, too.

BijuThomas-NOAA commented:
> I have this error message for regional_static_C192s1n4_atm_3denvar on Hercules and Orion, too.

I think it is a known issue and can be ignore for the time being.

JunghoonShin-NOAA commented:
On Ursa, all RT tests completed successfully except the "Regional standalone low-resolution static basin-focused configuration with 3DEnVar using GDAS ensembles" test.  
/scratch3/HFIP/hwrfv3/scrub/Junghoon.Shin/hafs.v2.1.1_rt_rt_regional_static_C192s1n4_atm_3denvar/2024092412/00L/hafs_analysis.log
The log file mentions this error message.
 ncdf error NetCDF: HDF error
 ncdf error NetCDF: HDF error
 ncdf error NetCDF: HDF error
 ncdf error NetCDF: HDF error
 ncdf error NetCDF: HDF error
 ncdf error NetCDF: HDF error
 ncdf error NetCDF: HDF error
 ncdf error NetCDF: HDF error

BijuThomas-NOAA commented:
RT were successful on C6(except regional_static_C192s1n4_atm_3denvar)
@BinLiu-NOAA we need to update the modulefiles (sorc/hafs_forecast.fd/modulefiles) for corresponds to C6/Orion/Herculs
I see that forceast is linked to support/HAFS in feature/hafs.v2.1.1

LinZhu-NOAA commented:
RT were successful on WCOSS2(except regional_static_C192s1n4_atm_3denvar).

ChuankaiWang-NOAA commented:
RT completed successfully on Orion and Hercules except for regional_static_C192s1n4_atm_3denvar.

BinLiu-NOAA committed:
Update submodule hafs_forecast.fd (addressing building/running issues on Orion/Hercules, from @BijuThomas-NOAA).
Turn off the RT of regional standalone low-resolution static basin-focused configuration with 3DEnVar using GDAS ensembles.
